src/plugins/explore/public/application/utils/state_management/middleware/query_sync_middleware.ts

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { Middleware } from '@reduxjs/toolkit';
+import { isEqual } from 'lodash';
+import { RootState } from '../store';
+import { ExploreServices } from '../../../../types';
+import { getQueryWithSource } from '../../languages';
+
+/**
+ * Middleware to sync Redux query state with global queryStringManager
+ * This ensures that whenever query state changes, the global state is updated
+ * before any search operations occur. It also manages query history for
+ * user-initiated query executions.
+ */
+export const createQuerySyncMiddleware = (services: ExploreServices): Middleware<{}, RootState> => {
+  return (store) => (next) => (action) => {
+    const result = next(action);
+
+    if (
+      action.type === 'query/setQueryState' ||
+      action.type === 'query/setQueryWithHistory' ||
+      action.type === 'query/setQueryStringWithHistory'
```

Comment:
wouldn't it be easier just have one? always require a query object and always populate it to the history?

i dont see why we would have three different types for this